sync/statesync/code_queue.go

```diff
@@ -0,0 +1,260 @@
+	errFailedToFinalizeCodeQueue    = errors.New("failed to finalize code queue")
+
+	// errAddCodeAfterFinalize is returned when [CodeQueue.AddCode] is called after [CodeQueue.Finalize] has been called.
+	// has permanently closed the queue to new work which guards against racy producers
+	// trying to enqueue during or after shutdown.
+	errAddCodeAfterFinalize = errors.New("cannot add code hashes after finalize")
+)
+
+// CodeQueue implements the producer side of code fetching.
+// It accepts code hashes, persists durable "to-fetch" markers (idempotent per hash),
+// and enqueues the hashes as-is onto an internal channel consumed by the code syncer.
+// The queue does not perform in-memory deduplication or local-code checks; the consumer
+// is responsible for deduping, skipping already-present code, and retrying.
+type CodeQueue struct {
+	db   ethdb.Database
+	quit <-chan struct{}
+
+	// Closed by [CodeQueue.Finalize] after the WaitGroup unblocks.
+	out       chan common.Hash
+	enqueueWG sync.WaitGroup
+
+	// Indicates why/if the output channel was closed.
+	closed atomicCloseState
+}
+
+// TODO: this will be migrated to using libevm's options pattern in a follow-up PR.
+type codeQueueOptions struct {
+	capacity int
+}
+
+type CodeQueueOption func(*codeQueueOptions)
+
+// WithCapacity overrides the queue buffer capacity.
+func WithCapacity(n int) CodeQueueOption {
+	return func(o *codeQueueOptions) {
+		if n > 0 {
+			o.capacity = n
+		}
+	}
+}
+
+// NewCodeQueue creates a new code queue applying optional functional options.
+func NewCodeQueue(db ethdb.Database, quit <-chan struct{}, opts ...CodeQueueOption) (*CodeQueue, error) {
+	// Apply defaults then options.
+	o := codeQueueOptions{
+		capacity: defaultQueueCapacity,
+	}
+	for _, opt := range opts {
+		opt(&o)
+	}
+
+	q := &CodeQueue{
+		db:   db,
+		out:  make(chan common.Hash, o.capacity),
+		quit: quit,
+	}
+
+	// Close the output channel on early shutdown to unblock consumers.
+	go func() {
+		<-q.quit
+		// If not already closed, mark as quit and close channel.
+		q.closed.markQuitAndClose(q.out)
+	}()
+
+	// Always initialize eagerly.
+	if err := q.init(); err != nil {
+		return nil, err
+	}
+
+	return q, nil
+}
+
+// CodeHashes returns the receive-only channel of code hashes to consume.
+func (q *CodeQueue) CodeHashes() <-chan common.Hash {
+	return q.out
+}
+
+// AddCode implements [syncpkg.CodeRequestQueue] by persisting and enqueueing new hashes.
+// Persists idempotent "to-fetch" markers for all inputs and enqueues them as-is.
+// Returns errAddCodeAfterFinalize after a clean finalize and errFailedToAddCodeHashesToQueue on early quit.
+func (q *CodeQueue) AddCode(codeHashes []common.Hash) error {
+	// If the queue has been closed, reject new work.
+	switch q.closed.get() {
+	case closeStateFinalized:
+		return errAddCodeAfterFinalize
+	case closeStateQuit:
+		return errFailedToAddCodeHashesToQueue
+	}
```

Comment:
```suggestion
```

It's not valid to call `AddCode()` after a call to `Finalized()` so the only way this can happen is if the `quit` channel is closed externally. In this case, `enqueue()` will return an error anyway and this `switch` won't guarantee avoiding it because it could race.

Reply:
It's unnecessary though because it's trying to manage improper use of the API contract, which is that it's invalid to call `AddCode()` after `Finalize()`. Trying to account for all the ways that consumers can do wrong only over-complicates the code here.

```go
switch q.closed.get() {
case closeStateFinalized:
    // The consumer has broken the API contract; it's not on this code to protect against the send-after-close panic.
case closeStateQuit:
    // enqueue() will return the same error anyway.
}
```